addons/website_event_online/models/event_event.py

```diff
@@ -3,12 +3,53 @@
 
 from odoo import models, _
 from odoo.addons.http_routing.models.ir_http import slug
+from odoo.osv import expression
 
 
 class Event(models.Model):
     _name = 'event.event'
     _inherit = 'event.event'
 
+    def _compute_is_participating(self):
+        """ Override is_participating to improve heuristic that is now
+
+          * public, no visitor: not participating as we have no information;
+          * public and visitor: check visitor is linked to a registration. As
+            visitors are merged on the top parent, current visitor check is
+            sufficient even for successive visits;
+          * logged, no visitor: check partner is linked to a registration. Do
+            not check the email as it is not really secure;
+          * logged as visitor: check partner or visitor are linked to a
+            registration;
+        """
+        current_visitor = self.env['website.visitor']._get_visitor_from_request(force_create=False)
+        if self.env.user._is_public() and not current_visitor:
+            self.is_participating = False
+        elif self.env.user._is_public():
+            participating = self.env['event.registration'].sudo().search([
+                ('event_id', 'in', self.ids),
+                ('visitor_id', '=', current_visitor.id),
+            ]).event_id
+            for event in self:
+                event.is_participating = event in participating
+        else:
+            if current_visitor:
+                domain = [
+                    '|',
+                    ('partner_id', '=', self.env.user.partner_id.id),
+                    ('visitor_id', '=', current_visitor.id)
+                ]
+            else:
+                domain = [('partner_id', '=', self.env.user.partner_id.id)]
+            participating = self.env['event.registration'].sudo().search(
+                expression.AND([
+                    domain,
+                    [('event_id', 'in', self.ids)]
+                ])
+            ).event_id
+            for event in self:
+                event.is_participating = event in participating
```

Comment:
Suggestion to simplify code logic a bit:

Could build a recordset of events the user is participating to.
Something like:

```py
if self.env.user._is_public() and not current_visitor:
    participating_events = self.env['event.event]
elif self.env.user._is_public()::
    participating_events = ...
else
    participating_events = ...

for event in self:
     event.is_participating = event in participating_events
```

So we have this last loop only once.

Reply:
Side note: This could be nicely tested in a python test to verify all different flows are working properly.

Reply:
Done (rewriting, python tests, well... "when we have time")